File: DataSets/DataSet-New/LeetCodeTasks - Develop/LeetCodeTasks/3077_maximum-strength-of-k-disjoint-subarrays/solutions/python3/solution.py
from typing import List
from functools import lru_cache

class Solution:
    def maximumStrength(self, nums: List[int], k: int) -> int:
        n = len(nums)
        prefix_sums = [0] * (n + 1)
        
        # Compute prefix sums
        for i in range(n):
            prefix_sums[i + 1] = prefix_sums[i] + nums[i]

        # Helper function to calculate sum of a subarray nums[l..r]
        def subarray_sum(l, r):
            return prefix_sums[r + 1] - prefix_sums[l]
        
        # Memoization for dynamic programming
        @lru_cache(None)
        def dp(i, rem_k):
            if rem_k == 0:
                return 0
            if i >= n:
                return float('-inf')
            
            max_strength = float('-inf')
            # Choose a subarray starting at index i
            for j in range(i, n):
                subarray_strength = subarray_sum(i, j)
                next_strength = dp(j + 1, rem_k - 1)
                # Calculate strength with current subarray included
                if next_strength != float('-inf'):
                    current_strength = (rem_k - (rem_k - 1)) * subarray_strength + next_strength
                    max_strength = max(max_strength, current_strength)
            return max_strength
        
        # Initialize dynamic programming approach starting from index 0 with k subarrays remaining
        return dp(0, k)

File: DataSets/DataSet-New/LeetCodeTasks - Develop/LeetCodeTasks/3077_maximum-strength-of-k-disjoint-subarrays/solutions/java/solution.java
class Solution {
    public long maximumStrength(int[] nums, int k) {
        int n = nums.length;
        long[] prefixSums = new long[n + 1];
        for (int i = 0; i < n; i++) {
            prefixSums[i + 1] = prefixSums[i] + nums[i];
        }

        // dp[i][j] will store the maximum strength using exactly j subarrays from first i elements
        long[][] dp = new long[n + 1][k + 1];
        
        // Initialize dp array for the case of using 0 subarrays (strength is 0)
        for (int i = 0; i <= n; i++) {
            for (int j = 0; j <= k; j++) {
                dp[i][j] = Long.MIN_VALUE;
            }
        }
        dp[0][0] = 0; // No subarrays, no elements, strength 0

        // Fill dp array
        for (int i = 1; i <= n; i++) {
            for (int j = 1; j <= Math.min(i, k); j++) {
                // We can only form j subarrays using at least j elements
                for (int m = 0; m < i; m++) {  // m is the starting index of the new subarray, m < i
                    long subarraySum = prefixSums[i] - prefixSums[m];
                    long weight = ((j % 2 == 1) ? 1 : -1) * (k - j + 1);
                    dp[i][j] = Math.max(dp[i][j], dp[m][j - 1] + weight * subarraySum);
                }
            }
        }

        return dp[n][k];
    }
}
